packages/firestore/src/local/indexeddb_query_cache.ts

```diff
@@ -97,36 +97,73 @@ export class IndexedDbQueryCache implements QueryCache {
     );
   }
 
-  addQueryData(
+  private saveQueryData(
     transaction: PersistenceTransaction,
     queryData: QueryData
   ): PersistencePromise<void> {
-    const targetId = queryData.targetId;
-    const addedQueryPromise = targetsStore(transaction).put(
-      this.serializer.toDbTarget(queryData)
-    );
-    if (targetId > this.metadata.highestTargetId) {
-      this.metadata.highestTargetId = targetId;
-      return addedQueryPromise.next(() =>
-        globalTargetStore(transaction).put(DbTargetGlobal.key, this.metadata)
-      );
-    } else {
-      return addedQueryPromise;
+    return targetsStore(transaction).put(this.serializer.toDbTarget(queryData));
+  }
+
+  private updateMetadataForQueryData(queryData: QueryData): boolean {
+    let needsUpdate = false;
+    if (queryData.targetId > this.metadata.highestTargetId) {
+      this.metadata.highestTargetId = queryData.targetId;
+      needsUpdate = true;
     }
+
+    // TODO(gsoltis): add sequence number check
+    return needsUpdate;
   }
 
-  removeQueryData(
+  private saveMetadata(
+    transaction: PersistenceTransaction
+  ): PersistencePromise<void> {
+    return globalTargetStore(transaction).put(
+      DbTargetGlobal.key,
+      this.metadata
+    );
+  }
+
+  addQueryData(
+    transaction: PersistenceTransaction,
+    queryData: QueryData
+  ): PersistencePromise<void> {
+    return this.saveQueryData(transaction, queryData).next(() => {
+      this.metadata.targetCount += 1;
+      this.updateMetadataForQueryData(queryData);
```

Comment:
I feel like 'updateMetadataForQueryData' should update all the metadata, including the target count.... and at that point it might as well persist to IndexedDB as well.

Reply:
We can't update the target count and share the logic between `addQueryTarget` and `updateQueryTarget`. One updates, the other adds, so the change to `targetCount` is different. We also can't tell which is which without doing a read or passing in a parameter, which I think would defeat the point.